### PR TITLE
[Performance] V1 Classify Models E2E Performance Optimization

### PR DESCRIPTION
**Pull request by noooop**
## Purpose

- Optimize the logic for the score task in supported_tasks
- V1 classify optimization

```
https://github.com/noooop/snippet/blob/main/benchmarks/classify/profile.py
VLLM_TORCH_PROFILER_DIR=/share/traces/ python -m benchmarks.classify.profile
```

Removing pooled_data.shape[-1] causes CUDA sync

8f4c9d8: CPU: 18.098ms CUDA: 17.271ms -> CPU: 18.163ms CUDA: 17.231ms

Very small improvement

cc @DarkLight1337 @maxdebayser

## (Optional) Documentation Update

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/layers/pooler.py [hunk 1]
```diff
@@ -389,19 +389,33 @@ def forward_chunk(self, pooled_data: torch.Tensor) -> torch.Tensor:
...
+
+        from vllm.config import get_current_vllm_config
+        vllm_config = get_current_vllm_config()
+        self.num_labels = getattr(vllm_config.model_config.hf_config,
+                                  "num_labels", 0)
+
     def forward_chunk(self, pooled_data: torch.Tensor) -> torch.Tensor:
-        num_labels = pooled_data.shape[-1]
-        if num_labels < 2:
+        if self.num_labels == 1:
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py]: How about the case where `num_labels == 0`?

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -1292,10 +1292,17 @@ def get_supported_pooling_tasks(self) -> list[PoolingTask]:
                 and "encode" in supported_tasks):
             supported_tasks.remove("encode")
 
-            logger.info_once("Chunked prefill is not supported with "
-                             "encode task which using ALL pooling. "
-                             "Please turn off chunked prefill by "
-                             "`--no-enable-chunked-prefill` before using it.")
+            logger.debug_once("Chunked prefill is not supported with "
```
Comment (DarkLight1337) [inline vllm/v1/worker/gpu_model_runner.py:1251]: Why change to debug?

File: vllm/model_executor/layers/pooler.py (hunk 1 above)
Comment (noooop) [reply vllm/model_executor/layers/pooler.py]: num_labels is theoretically impossible to be 0, so 0 is for fallback to softmax. e.g., supporting reward models

Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py]: Can you log a warning in the initialization if `num_labels` is found to be 0?

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
Comment (noooop) [reply vllm/v1/worker/gpu_model_runner.py:1251]: Is info_once a bit verbose?

Comment (DarkLight1337) [reply vllm/v1/worker/gpu_model_runner.py:1251]: The default logging level is INFO so this message won't be shown at all to most users. Is that what you want?

File: vllm/model_executor/layers/pooler.py (hunk 1 above)
Comment (noooop) [reply vllm/model_executor/layers/pooler.py]: Reward models also use PoolerClassify, theoretically it does not set the num_labels parameter, possibly num_labels == 2, or num_labels == 0.

I don't think there's a big problem with falling back to softmax

***

To be honest, I don't know whether reward models or decode tasks would have corner cases (e.g., dimension= 0 or dimension= 1)  if we don't use pooled_data.shape[-1] to dynamically get the dimension.

Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py]: This changes the behavior from before though, since previously we apply sigmoid instead of softmax if `num_labels == 0`... perhaps we should instead check the dimensions of the tensor?

File: vllm/model_executor/layers/pooler.py [hunk 3]
```diff
@@ -396,6 +399,10 @@ def __init__(self):
         vllm_config = get_current_vllm_config()
         self.num_labels = getattr(vllm_config.model_config.hf_config,
                                   "num_labels", 0)
+        if self.num_labels == 0:
+            logger.warning("It's hard to imagine num_labels == 0, "
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py]:
```suggestion
            logger.warning("num_labels should be > 0 for classification models, "
```

File: vllm/model_executor/layers/pooler.py [hunk 4]
```diff
@@ -414,6 +421,11 @@ def __init__(self):
         self.num_labels = getattr(vllm_config.model_config.hf_config,
                                   "num_labels", 0)
 
+        if self.num_labels == 0:
+            logger.warning("It's hard to imagine num_labels == 0, "
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py]: Similar idea here

Comment (noooop) [reply vllm/model_executor/layers/pooler.py]: I'm a bit busy tonight, I'll update tomorrow.

File: vllm/model_executor/layers/pooler.py [hunk 5]
```diff
@@ -661,13 +686,8 @@ def forward(
         # pooled_data shape: [batchsize, hidden_size]
 
         if self.classifier is not None:
-            # apply classifier once on the full batch if possible
-            if isinstance(pooled_data, torch.Tensor):
-                pooled_data = self.classifier(pooled_data)
-            elif len({data.shape for data in pooled_data}) <= 1:
-                pooled_data = self.classifier(torch.stack(pooled_data))
-            else:
-                pooled_data = [self.classifier(data) for data in pooled_data]
```
Comment (maxdebayser) [inline vllm/model_executor/layers/pooler.py:652]: In theory it's possible for the pooled data to be a list of tensors. For example, if the the pooler is AllPool. I don't know if such a combination exists, but if yes it would cause a crash here.

Comment (noooop) [reply vllm/model_executor/layers/pooler.py:652]: https://github.com/vllm-project/vllm/blob/6fad29b11b3680c44782cd6e5fe555779d620d6c/vllm/model_executor/layers/pooler.py#L288-L292


AllPool does not support "embed"/ "classify"/ "score", let's make things simpler.

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
Comment (noooop) [reply vllm/v1/worker/gpu_model_runner.py:1251]: The most common situation ordinary users encounter is

```
vllm serve jason9693/Qwen2.5-1.5B-apeach
# Supported_tasks: ['classify']
```

Disabling the "encode" and "score" api will almost not make users lose anything, but explaining to users will cause them to be more confused.

So I think this logging is best only shown when in debug mode

**Review by maxdebayser — APPROVED** [review]
LGTM

**Comment by noooop** [comment]
@DarkLight1337 

Is there anything else that needs to be modified in this PR?

File: vllm/model_executor/layers/pooler.py [hunk 6]
```diff
@@ -371,22 +374,29 @@ def forward_chunk(self, pooled_data: torch.Tensor) -> torch.Tensor:
...
 class PoolerClassify(PoolerActivation):
 
-    def forward_chunk(self, pooled_data: torch.Tensor) -> torch.Tensor:
-        num_labels = pooled_data.shape[-1]
-        if num_labels < 2:
-            return F.sigmoid(pooled_data.float()).to(pooled_data.dtype)
-
-        return F.softmax(pooled_data.float(), dim=-1).to(pooled_data.dtype)
-
+    def __init__(self, static_num_labels=True):
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py]:
```suggestion
    def __init__(self, *, static_num_labels: bool = True) -> None:
```

Add type annotation

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
Comment (DarkLight1337) [reply vllm/v1/worker/gpu_model_runner.py:1251]: Ok

**Review by DarkLight1337 — APPROVED** [review]

**Comment by noooop** [comment]
@DarkLight1337

flaky test
